ibtk/src/utilities/HierarchyIntegrator.cpp

```diff
@@ -501,15 +486,15 @@ HierarchyIntegrator::regridHierarchy()
         TBOX_WARNING(
             d_object_name << "::regridHierarchy():\n"
                           << "  change in domain volume detected (volume is computed by summing control volumes)\n"
-                          << "    old volume = "
-                          << old_volume
-                          << "\n"
-                          << "    new volume = "
-                          << new_volume
-                          << "\n"
+                          << "    old volume = " << old_volume << "\n"
+                          << "    new volume = " << new_volume << "\n"
                           << "  this may indicate overlapping patches in the AMR grid hierarchy.");
     }
 
+    // Reinitialize composite grid data.
+    const bool initial_time = false;
+    initializeCompositeHierarchyData(d_integrator_time, initial_time);
+
```

Comment:
And also from here. The top integrator makes this call explicitly.

Reply:
Same comment as above. I don't want to inject this into all of the implementations.

Reply:
Here is a correct sequence:

https://github.com/IBAMR/IBAMR/blob/amneetb/multiphase-flows/src/IB/IBHierarchyIntegrator.cpp#L403-L427

I would also love to do this automatically, but now sure how.

Reply:
I prefer the name compositeHierarchy(), in which the intent is clear--- fill data only when entire hierarchy has been created. The level-by-level filling is tagging and creating levels on the fly and hierarchy is far from being done.

Reply:
Could you just do the LData initialization in the initializeCompositHierarchyData() class implementation?

Reply:
E.g. IBHierarchyIntegrator::initializeCompositeHierarchyDataSpecialized() does the data redistribution?

Reply:
You mean application specific or IBHierachyIntegrator specific? 

LData is created level-by-level basis, and cells are tagged on the level and next level created. It does not seem like a compositeHierarchyData() problem. You of course are refering to something else that I am not getting, because this is how you designed.

Reply:
One can potentially stick LData-->begin/endRedistribution() into IBHierarchyIntegrator::initializeCompositeHierarchyDataSpecialized(), but I think LData is fundamentally a level-by-level init problem, and not a compositeHierarchyInit problem. The former plays in role in deciding how the hierarchy shapes up, while the later is helpless in that regard.

Reply:
Sorry, I meant Lagrangian data redistribution --- not LData initialization.

Looking at the implementation, it seems like you should be able to put the redistribution in the initializeCompositeHierarchyDataSpecialized() method, and then it will be ready to use for any Eulerian data setting that relies on the updated distributions being available.

Reply:
Hmm, OK. If redistribution is just relying on existing hierarchy then it is OK. I have not looked too carefully into that routine.

Reply:
At the initial time there is only one patch hierarchy to deal with. :-)

It is at later times, when the hierarchy actually changes, that you need to split up the redistribution stuff into two phases.

Reply:
You would need to only do the data redistribution at initial time, since other times the redistribution "straddles" regridding.

Reply:
Yeah, that was my impression too. LData-->begin/endRedistribution can alter the hierarchy itself. The routine initCompositeHierarchy() means the hierarchy is not disrupted. But I guess the IBHierarchyIntegrator::initCompositeHierSpecialized() knows that to do these things in a specific order. 

I can stick begin/end dataRedistribution in the IBHierarchyIntegrator::initCompositeHierSpecialized() method.

Reply:
OK, great.

With that change, I think that the interface set up in this PR should work in the other branch too. I will merge once Jenkins is happy.

Reply:
Well, I am confused. The current sequence now is 

1) HierarchyIntegrator::initPatchHier
2) LData-->begin/end
3) HierarchyIntegrator::InitComposite


Now it would be 
HierarchyIntegrator::initPatchHier-->which calls HierarchyIntegrator::InitCompositeSpecialized()

How would HierarchyIntegrator::InitCompositeSpecialized() look without being recurcive?